lib/ssl/src/ssl_gen_statem.erl

```diff
@@ -1345,13 +1352,36 @@ handle_sni_hostname(Hostname,
     end.
 
 update_ssl_options_from_sni(#{sni_fun := SNIFun} = OrigSSLOptions, SNIHostname) ->
-    case SNIFun(SNIHostname) of
+    try SNIFun(SNIHostname) of
         undefined ->
-            undefined;
+            use_original_options;
+        unrecognized ->
+            ?ALERT_REC(?FATAL, ?UNRECOGNIZED_NAME, {sni, SNIHostname});
         SSLOptions ->
-            VersionsOpt = proplists:get_value(versions, SSLOptions, []),
-            FallBackOptions = filter_for_versions(VersionsOpt, OrigSSLOptions),
-            ssl:update_options(SSLOptions, server, FallBackOptions)
+            try
+                FallBackOptions = fallback_options(SSLOptions, OrigSSLOptions),
+                ssl:update_options(SSLOptions, server, FallBackOptions)
+            catch
+                throw:#alert{} = Alert ->
+                    Alert;
+                throw:Reason ->
+                    ?LOG_ERROR("sni_fun provided erroneous options : ~p~n", [Reason]),
+                    ?ALERT_REC(?FATAL, ?HANDSHAKE_FAILURE, sni_fun_provided_erroneous_options)
+            end
+    catch
+        Error:Reason ->
+            ?LOG_ERROR("sni_fun crashed ~p ~p~n", [Error, Reason]),
+            ?ALERT_REC(?FATAL, ?HANDSHAKE_FAILURE, sni_fun_crashed)
+    end.
+
+fallback_options(SSLOptions, OrigSSLOptions) ->
+    VersionsOpt = proplists:get_value(versions, SSLOptions, []),
+    try filter_for_versions(VersionsOpt, OrigSSLOptions) of
+        FallBackOptions ->
+            FallBackOptions
+    catch _:_ ->
+            ?LOG_ERROR("sni_fun provided erroneous options : ~p~n", [VersionsOpt]),
```

Comment:
Include error and stacktrace in log_error

Reply:
fixed